rir/src/interpreter/interp.cpp

```diff
@@ -123,6 +123,50 @@ static void jit(SEXP cls, Context* ctx) {
     SET_BODY(cls, BODY(cmp));
 }
 
+void tryOptimizeClosure(DispatchTable* table, size_t offset, Context* ctx) {
+
+    // This function will optimize the function and update the linked list
+    // of functions, so that subsequent closure calls pick the most optimized
+    // version.
+
+    static bool optimizing = false;
+
+    Function* fun = table->at(offset);
+
+    if (!optimizing && !fun->next() && !fun->origin()) {
+         /* currently there is a bug if we reoptimize a function twice:
+          *  Deopt ids from the first optimization and second optimizations
+          *  will be mixed and the deoptimizer always only goes back one
+          *  optimization level!
+          *  To avoid this bug we currently only optimize once
+          */
+        Code* code = fun->body();
+        if (fun->markOpt ||
+                (fun->invocationCount == 1 && code->perfCounter > 100) ||
+                (fun->invocationCount == 10 && code->perfCounter > 20) ||
+                 fun->invocationCount == 100) {
+            optimizing = true;
+
+            Function* oldFun = fun;
+            cp_pool_add(ctx, oldFun->container());
+
+            SEXP opt = globalContext()->optimizer(fun->container());
+
+            if (opt != nullptr) {
+                fun = Function::unpack(opt);
```

Comment:
There used to be a `PROTECT` and `UNPROTECT` here. Was it removed because we're already adding the container to the constant pool (line 151)?

Reply:
No, the `PROTECT` was for the container of the newly optimized version. However, I think it is not needed: `vtable->put()` should not cause a GC, since it does not reallocate the table - it just writes to the same slot. And what happens inside is just a write barrier check, and that should not collect either, right?